src/runtime/micro/micro_session.h

```diff
@@ -0,0 +1,244 @@
+   */
+  const char* type_key() const final {
+    return "MicroSession";
+  }
+
+  /*!
+   * \brief constructor
+   */
+  MicroSession();
+
+  /*!
+   * \brief destructor
+   */
+  ~MicroSession();
+
+  static std::shared_ptr<MicroSession>& Current();
+
+  /*!
+   * \brief creates session by setting up a low-level device and initting allocators for it
+   * \param args TVMArgs passed into the micro.init packedfunc
+   */
+  void CreateSession(const std::string& device_type,
+                     const std::string& binary_path,
+                     const std::string& toolchain_prefix);
+
+  /*!
+   * \brief ends the session by destructing the low-level device and its allocators
+   */
+  void EndSession();
+
+  /*!
+   * \brief allocate memory in section
+   * \param type type of section to allocate in
+   * \param size size of allocated memory in bytes
+   * \return pointer to allocated memory region in section, nullptr if out of space
+   */
+  DevBaseOffset AllocateInSection(SectionKind type, size_t size);
+
+  /*!
+   * \brief free prior allocation from section
+   * \param type type of section to allocate in
+   * \param ptr pointer to allocated memory
+   */
+  void FreeInSection(SectionKind type, DevBaseOffset ptr);
+
+  /*!
+   * \brief read string from device to host
+   * \param str_offset device offset of first character of string
+   * \return host copy of device string that was read
+   */
+  std::string ReadString(DevBaseOffset str_offset);
+
+  /*!
+   * \brief sets up init stub pointers and copies arguments for on-device execution
+   * \param func address of the function to be executed
+   * \param args args to the packed function
+   */
+  void PushToExecQueue(DevBaseOffset func, const TVMArgs& args);
+
+  /*!
+   * \brief loads binary onto device
+   * \param binary_path path to binary object file
+   * \return info about loaded binary
+   */
+  BinaryInfo LoadBinary(std::string binary_path);
+
+  /*!
+  * \brief read value of symbol from device memory
+  * \param symbol_map symbol map to read location of symbol from
+  * \param symbol name of symbol being read from
+  * \return value at symbol in memory
+  */
+  template <typename T>
+  T DevSymbolRead(const SymbolMap& symbol_map, const std::string& symbol);
+
+  /*!
+  * \brief write value into device memory corresponding to symbol
+  * \param symbol_map symbol map to read location of symbol from
+  * \param symbol name of symbol being written to
+  * \param value value being written into symbol
+   */
+  template <typename T>
+  void DevSymbolWrite(const SymbolMap& symbol_map, const std::string& symbol, const T& value);
+
+  /*!
+   * \brief returns low-level device pointer
+   * \note assumes low-level device has been initialized
+   */
+  const std::shared_ptr<LowLevelDevice> low_level_device() const {
+    CHECK(low_level_device_ != nullptr) << "attempt to get uninitialized low-level device";
+    return low_level_device_;
+  }
+
+  SymbolMap& init_symbol_map() {
```

Comment:
is this only used by member function? if so, move it to private. It is a bit uncommon to expose a non-const reference member

Reply:
It's used [here](https://github.com/dmlc/tvm/pull/3227/files#diff-a017c34eeae79dcd77b0b23e6c3fd009R97), but now that I think about it, `PatchImplHole` should be in `MicroSession` and not in `MicroModule`.  I'll move it, then make `init_symbol_map` private.